### PR TITLE
Fairmandering Project Review PR

### PR DESCRIPTION
**Pull request by samar-khanna**
Hello @fairmandering here is your review branch.

### CODE REVIEW

File: cora/comparison_map.html [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    href="https://api.mapbox.com/mapbox-gl-js/plugins/mapbox-gl-compare/v0.4.0/mapbox-gl-compare.css" type="text/css" />
+  <div id="comparison-container">
+    <div id="before" class="map"></div>
+    <div id="after" class="map"></div>
+  </div>
+  <script>
+    // TO MAKE THE MAP APPEAR YOU MUST
+    // ADD YOUR ACCESS TOKEN FROM
+    // https://account.mapbox.com
+    mapboxgl.accessToken = 'pk.eyJ1IjoiZm9yYWdlciIsImEiOiJjazVlNTV6ZmwxemwwM2twbzVocWljZjR1In0.g7pCpemtWzpHN6WEXMnFUw';
```
Comment (samar-khanna) [inline cora/comparison_map.html:59]: Implementation suggestion: You should avoid putting raw access tokens for any user in the source code. If someone decides to `inspect source` you might get into trouble.

File: cora/comparison_map.html [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      container: 'before',
+      style: 'mapbox://styles/forager/ck5e8bzry0fkd1iqmofttjes4',
+      center: [-80, 35.2],
+      maxBounds: [
+        [-86, 33],
+        [-74, 37]
+      ]
+    });
+
+    beforeMap.on('load', function () {
```
Comment (samar-khanna) [inline cora/comparison_map.html:75]: Documentation suggestion: It would be useful to understand what `beforeMap` and `afterMap` are supposed to represent. Please add some comments/documentation explicating this.

File: cora/generate_svg.html [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  width = divWidth;
+  height = divWidth;
+  console.log(width)
+  const margin = { top: 20, right: 20, bottom: 20, left: 20 };
+  chartWidth = width - margin.left - margin.right;
+  chartHeight = height - margin.top - margin.bottom;
+
+  var arr = [];
+  var n = 0;
+  for (var r = 0; r < 9; r++) {
```
Comment (samar-khanna) [inline cora/generate_svg.html:43]: Documentation suggestion: A cursory glance is telling me that this is generating a 9x9 array displaying some objects. Could you add comments explaining what is being created?

File: cora/website/index.html [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+            Fairmandering</p>
+    </footer>
+
+</body>
+
+<script src="static/js/flickity.pkgd.min.js"></script>
+<script>
+    (function ($) {
+
+        /**
```
Comment (samar-khanna) [inline cora/website/index.html:350]: Thank you for putting the documentation here!

File: evelyn/misc/precincts.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
+from gerrypy import constants
```
Comment (samar-khanna) [inline evelyn/misc/precincts.py:1]: Documentation suggestion: Very neat code, but difficult to understand without function doc strings. Also @ew423 could you tell me what you implemented in this file, am curious to know.

File: evelyn/misc/precincts.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+                mock = np.empty(len(election_columns))
+                mock[:] = np.nan
+                tract_election_results[t] = pd.Series(mock)
+
+        col_names = {ix: estr for ix, estr in enumerate(election_columns)}
+        tract_election_df = pd.DataFrame(tract_election_results).T.rename(columns=col_names).fillna(0)
+
+        return tract_election_df
+
+    def infer_w_county_data(self, votes_df):
```
Comment (samar-khanna) [inline evelyn/misc/precincts.py:131]: Structure suggestion: This function looks very dense. Am not sure what those hard-coded indices represent.

File: evelyn/misc/shapefiles.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
+"""This script downloads shapefiles from all 50 states and American territories
```
Comment (samar-khanna) [inline evelyn/misc/shapefiles.py:1]: Thank you for the high-level description for this script!

File: evelyn/misc/shapefiles.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    year = str(year)
+
+    try:
+        os.mkdir(shapes_dir)
+    except FileExistsError:
+        pass
+
+    # Check if already have
+    downloaded_files = os.listdir(shapes_dir)
+    downloaded_files = [f.split("_") for f in downloaded_files]
```
Comment (samar-khanna) [inline evelyn/misc/shapefiles.py:44]: Implementation suggestion: You might not need it, but it could be safer to check that after `os.listdir(..)`, you filter for the files with the right extension/ name pattern. Usually not an issue, but I have faced some nefarious and hard to find bugs when I forgot to do this in some projects.

File: winnie/gerrypy/data/load.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
+import pickle
```
Comment (samar-khanna) [inline winnie/gerrypy/data/load.py:1]: Documentation suggestion: Please document all your functions. That should include what the function does, and what the important parameters are.

File: winnie/gerrypy/data/load.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    if os.path.exists(os.path.join(data_base_path, 'lengths.npy')):
+        lengths_path = os.path.join(data_base_path, 'lengths.npy')
+        lengths = np.load(lengths_path)
+    else:
+        from scipy.spatial.distance import pdist, squareform
+        lengths = squareform(pdist(state_df[['x', 'y']].values))
+
+    if os.path.exists(os.path.join(data_base_path, 'edge_dists.p')):
+        edge_dists_path = os.path.join(data_base_path, 'edge_dists.p')
+        edge_dists = pickle.load(open(edge_dists_path, 'rb'))
```
Comment (samar-khanna) [inline winnie/gerrypy/data/load.py:78]: Implementation suggestion: It's always better to use the Pythonic `with` syntax for file descriptors:
```suggestion
        with open(edge_dists_path, 'rb') as f:
            edge_dists = pickle.load(f)
```

File: winnie/local_search.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            obj[d] = self.obj_func(tracts)
+        print(obj)
+
+        # Check all the edges to identify adj nodes in two district, switch them to get a new solution
+        solution = copy.deepcopy(initial_solution)
+        for u, v in self.G.edges():
+            if district_of_node[u] != district_of_node[v]: 
+                old_d1 = district_of_node[u]
+                new_tracts1 = copy.deepcopy(solution[old_d1])
+                new_tracts1.remove(u)
```
Comment (samar-khanna) [inline winnie/local_search.py:68]: **Implementation suggestion**: I'm not sure if order matters or not, but `remove(..)` is an `O(n)` operation and you could greatly speed things up if you used sets instead of lists (for sets, `remove` is `O(1)`). This is of course only if order doesn't matter, which, for the purposes of this hill climbing implementation, might be the case.

File: winnie/local_search.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        for d, tracts in initial_solution.items():
+            obj[d] = self.obj_func(tracts)
+        print(obj)
+
+        # Check all the edges to identify adj nodes in two district, switch them to get a new solution
+        solution = copy.deepcopy(initial_solution)
+        for u, v in self.G.edges():
+            if district_of_node[u] != district_of_node[v]: 
+                old_d1 = district_of_node[u]
+                new_tracts1 = copy.deepcopy(solution[old_d1])
```
Comment (samar-khanna) [inline winnie/local_search.py:67]: Implementation suggestion: Why do you need to `deepcopy` the tracts each time? It seems for each hill climbing run, you are anyways `deepcopy`ing the `initial_solution`. The only place where I can see this affecting anything is if the new objective is actually not larger than the previous objective. For this I'd suggest using an `else` statement to revert the changes made to the tracts for `d1` and `d2`, which if you are using sets will be `O(1)`. Or you can use operations that create temporary new lists for `tract1` and `tract2` (which might still be `O(n)`, but will be faster than `deepcopy` if your data structures are large) and only assign these new lists to the solution if the objective is better. Of course, I do not know the exact intricacies of the data structures you are using, so this is just food for thought.

File: winnie/local_search.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                new_tracts1.remove(u)
+                new_tracts1.append(v)
+                old_d2 = district_of_node[v]
+                new_tracts2 = copy.deepcopy(solution[old_d2])
+                new_tracts2.remove(v)
+                new_tracts2.append(u)
+                # Test whether it improves the objective
+                new_obj1, new_obj2 = self.obj_func(new_tracts1), self.obj_func(new_tracts2)
+                if (new_obj1 + new_obj2) < (obj[old_d1] + obj[old_d2]):
```
Comment (samar-khanna) [inline winnie/local_search.py:76]: Implementation suggestion: Since you are just changing one node in the tracts before computing the objective, is there a way to cache objective values for the tracts? Also, is there a way to optimise your objective functions such that if you know the one node that is being changed between a previous computation and this computation, you can calculate the new objective value without having to go through all nodes in the tract?

File: winnie/local_search.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                    solution[old_d1], solution[old_d2] = new_tracts1, new_tracts2
+                    obj[old_d1], obj[old_d2] = new_obj1, new_obj2 # update with values of new solution so as to find the optimum
+                    district_of_node[u], district_of_node[v] = old_d2, old_d1 # update district of the node
+                    
+        return solution
+
+    def is_feasible(self, solution):
```
Comment (samar-khanna) [inline winnie/local_search.py:83]: Structure suggestion: I don't see this function used anywhere? Is it supposed to be used with the hill climbing?

File: winnie/local_search.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+                return False
+
+        return True
+
+def dispersion(tracts):
+    ''' Given a district (list of tracts), returns the dispersion '''
+    population = state_df.loc[tracts]['population'].values
+    dlocs = state_df.loc[tracts][['x', 'y']].values
+    centroid = np.average(dlocs, weights=population, axis=0)
+    geo_dispersion = (np.subtract(dlocs, centroid) ** 2).sum(axis=1) ** .5 / 1000
```
Comment (samar-khanna) [inline winnie/local_search.py:107]: Implementation suggestion: Why `np.subtract` and not just `-`?

File: winnie/local_search.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    dispersion = np.average(geo_dispersion, weights=population)
+    return dispersion
+
+if __name__ == "__main__":
```
Comment (samar-khanna) [inline winnie/local_search.py:111]: Structure suggestion: You might want to add experiments in a different file, better practice to do so.

File: evelyn/misc/precinct_state_wrappers.py [hunk 17]
```diff
@@ -0,0 +1,1328 @@
+import os
```
Comment (samar-khanna) [inline evelyn/misc/precinct_state_wrappers.py:1]: This seems like a template file, so I don't have too much to say. Maybe just a little bit of documentation for the file explaining what it is doing?

**Review by samar-khanna — COMMENTED** [review]
Please remove the images. Other than that, the code is pretty structured, some files lack documentation but the general organisation is quite professional. Please do add documentation to explain what the functions you wrote are doing, and if there are pieces of @wesg52 's code that aren't documented, it might be a good exercise for you guys to document some of it to understand more thoroughly what is going on. 

I liked that one of the files had a brief docstring about what the script's intended functionality is. That could be a good way to keep track of what each script does.

I can't comment too much on the web-side of things, so for that I just have documentation suggestions.

File: winnie/local_search.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
+from gerrypy.data.load import load_opt_data
+import networkx as nx
+import matplotlib.pyplot as plt
+import copy
+import numpy as np
+from scipy.spatial import distance
+import time
+
+'''
```
Comment (samar-khanna) [inline winnie/local_search.py:9]: What is this block of docstring?

File: felix/dissimilarity.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    the number of districts, and the third dimension a binary array of length (num_blocks),
+    where a 1 in the ith entry indicates that the district contains block i.
+    Args:
+        plans: the list of maps to convert
+        num_blocks: the largest number of blocks in a single district. Used for the conversion
+        to binary arrays.
+    """
+    possible_districts = pd.Series(np.arange(0, num_blocks))
+    converted_plans = []
+    for i in range(len(plans)):
```
Comment (samar-khanna) [inline felix/dissimilarity.py:38]: Implementation suggestion: if you want to be succinct, you could shorten this to:
```
possible_districts = pd.Series(np.arange(0, num_blocks))
converted_plans = [[possible_districts.isin(district) for district in plan] for plan in plans]
return np.array(converted_plans, dtype=np.int)
```

File: felix/dissimilarity.py [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+            dist_matrix[i][j] = dist
+            dist_matrix[j][i] = dist
+        count += 1
+        if count % dimension == 0:
+            i += 1
+        j = count % dimension
+
+    return dist_matrix
+
+def two_map_distance(map_1, map_2):
```
Comment (samar-khanna) [inline felix/dissimilarity.py:73]: Documentation suggestion: What kind of distance is this? Also, by itself, this distance is not symmetric (i.e. if map2 and map1 were flipped). Nor is the distance between `map1` and `map2` zero if `map1 == map2`

File: felix/dissimilarity.py [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+        plan = plans[i]
+        converted_plan = []
+        for j in range(len(plan)):
+            district = plan[j]
+            converted_plan.append(possible_districts.isin(district))
+        converted_plans.append(converted_plan)
+    return np.array(converted_plans).astype(int)
+
+
+def compute_dist_matrix(maps):
```
Comment (samar-khanna) [inline felix/dissimilarity.py:48]: Implementation suggestion: This might be faster (no loops). You could omit the line adding the symmetry if you don't care about symmetry, or if you are careful enough to only use indices for `i >= j`
```suggestion
def compute_dist_matrix(maps):
    # Compute matrix multiplication between each pair of maps, transpose it to (maps, maps, districts, blocks)
    prod = np.tensordot(maps, maps.T, axes=1).transpose(0, 3, 1, 2) 
    dist = 1000 * (1/np.mean(np.max(prod, axis=-1), axis=-1)) 
    # To make it symmetric 
    dist[np.triu_indices(dist.shape[0], 1)] = dist.T[np.triu_indices(dist.shape[0], 1)] 
    return dist 
```

File: felix/dissimilarity.py [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+    Return the distance between map_1 and map_2
+    Args:
+        map_1: (ndarray) 2 Dimensional ndarray representing a map
+        map_2: (ndarray) 2 Dimensional ndarray representing a map
+    """
+    new_map = map_1 @ map_2
+    return 1000 * (1 / np.mean(np.max(new_map, 1)))
+
+
+def furthest_maps(dist_matrix, maps, num_maps):
```
Comment (samar-khanna) [inline felix/dissimilarity.py:84]: Implementation suggestion: Here is my version of this function, keeping the same intended greedy algorithmic functionality :
```suggestion
def farthest_maps(dist, maps, k):
    output_indices = []
    output_indices.extend(np.unravel_index(np.argmax(dist), dist.shape))

    for i in range(2, k):
        dist_to_out = dist[output_indices, :]
        dist_to_out[:, output_indices] = 0.0

        sum_dist_to_out = np.sum(dist_to_out, axis=0)
        output_indices.append(np.argmax(sum_dist_to_out))
    
    output_maps = maps[output_indices, ...]
    assert (len(np.unique(output_maps, axis=0)) == k), "Re-using previously found maps"
    return output_maps
```

File: felix/dissimilarity.py [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+    """
+    dist = 0
+    if elem in output_indices:
+        return dist
+    else:
+        for i in range(0, len(output_indices)):
+            dist += dist_matrix[output_indices[i], elem]
+        return dist
+
+def convert_back(plans):
```
Comment (samar-khanna) [inline felix/dissimilarity.py:139]: Implementation suggestion: I think there is a way to limit the use of loops for this function `convert_back` as well, but I'm going to let you figure it out.

File: felix/dissimilarity.py [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+    num_plans = plans.shape[0]
+    num_districts = plans.shape[1]
+    converted_plans = []
+    for i in range(num_plans):
+        converted_plan = []
+        plan = plans[i]
+        for j in range(num_districts):
+            district = plan[j]
+            district = np.where(district)[0]
+            district[district != 0]
```
Comment (samar-khanna) [inline felix/dissimilarity.py:152]: Implementation suggestion: this line `district[district != 0]` is a no-op

File: felix/dissimilarity.py [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+            converted_plan.append(district.tolist())
+        converted_plans.append(converted_plan)
+    return converted_plans
+
+
+def compute_difference(dist_matrix, selected_maps):
```
Comment (samar-khanna) [inline felix/dissimilarity.py:158]: Structure suggestion: I don't see this function being used anywhere.

File: felix/graph_clustering.ipynb [hunk 26]
```diff
@@ -0,0 +1,443 @@
+{
```
Comment (samar-khanna) [inline felix/graph_clustering.ipynb:1]: Since this is your experiment file, I'm not going to go through it exhaustively.

File: felix/graph_clustering.ipynb [hunk 27]
```diff
@@ -0,0 +1,443 @@
...
+    "            clusters[np.argmin(dist_to_median)].append(graph)\n",
+    "        \n",
+    "        isIdentical = True \n",
+    "        for i in range(0, k): \n",
+    "            if medians[i] != prev_medians[i]:\n",
+    "                isIdentical = False\n",
+    "        if isIdentical == True: \n",
+    "            return medians\n",
+    "\n",
+    "def graph_dist_matrix(graph_maps): \n",
```
Comment (samar-khanna) [inline felix/graph_clustering.ipynb:150]: Implementation suggestion: Refer to my edits for the `dist_matrix` function to vectorise this.

File: felix/graph_clustering.ipynb [hunk 28]
```diff
@@ -0,0 +1,443 @@
...
+    "### Spectral Clustering representing a plan as a node in graph"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def align_plans(plans): \n",
```
Comment (samar-khanna) [inline felix/graph_clustering.ipynb:203]: Implementation suggestion: I'm kinda tired, so not going to go through this extensively. But if this is not a preprocessing function (i.e. you need to call it multiple times) find a way to vectorise it.

**Review by samar-khanna — COMMENTED** [review]
I like the documentation. Makes reviewing much easier. Check my edits for implementation/efficiency, and maybe @wesg52 can also guide you here. 

But I'm unhappy with how late you have submitted this code.

File: evelyn/misc/precinct_state_wrappers.py [hunk 29]
```diff
@@ -0,0 +1,1328 @@
...
+from gerrypy import constants
+from gerrypy.data.precincts import StatePrecinctWrapper
+
+
+# TODO: add validation sources where available
+
```
Comment (ps789) [inline evelyn/misc/precinct_state_wrappers.py:7]: Is there a way to simplify this without having wrapper classes for each state?

File: evelyn/misc/precincts.py [hunk 30]
```diff
@@ -0,0 +1,213 @@
+from gerrypy import constants
+from gerrypy.data.load import *
```
Comment (ps789) [inline evelyn/misc/precincts.py:2]: Would use comments describing the functions and what it's taking in, returning, etc

File: winnie/gerrypy/data/load.py [hunk 31]
```diff
@@ -0,0 +1,82 @@
+import pickle
+from gerrypy import constants
```
Comment (ps789) [inline winnie/gerrypy/data/load.py:2]: function details such as parameters and what it returns in what format is important

File: evelyn/misc/shapefiles.py [hunk 32]
```diff
@@ -0,0 +1,127 @@
+"""This script downloads shapefiles from all 50 states and American territories
+ from the TIGER census geospatial database. Shapefiles are saved in a folder
+based on their respective state codes as defined by:
+https://www2.census.gov/geo/docs/reference/state.txt?#
+
+Simply run <python get_shapefile.py> in the directory you wish to store
```
Comment (yangkt) [inline evelyn/misc/shapefiles.py:6]: Don't think there is a get_shapefile.py file atm

File: evelyn/misc/shapefiles.py [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+
+def download_state_shapes(states=None, year=constants.ACS_BASE_YEAR):
+    """Download census shapefiles for states.
+    
+    states: str list - state abbreviations to download
+    year : str - year for data to be collected"""
+    dirname = os.path.dirname(__file__)
+    shapes_dir = os.path.join(dirname, "shapes")
+    year = str(year)
+    TIGER_URL = "https://www2.census.gov/geo/tiger/TIGER%s/TRACT/" % year
```
Comment (yangkt) [inline evelyn/misc/shapefiles.py:27]: Would it be possible to make the url without the year global? seem to be assigning this variable multiple time in various functions.

File: cora/generate_svg.html [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+  <div class="reps-div">
+    <h4>Representatives</h4>
+    <i id="rep1" class="fa fa-user fa-7x"></i>
+    <i id="rep2" class="fa fa-user fa-7x"></i>
+    <i id="rep3" class="fa fa-user fa-7x"></i>
+  </div>
+</body>
+
+<script>
+  // Change steps and uncomment code at the bottom to download svg
```
Comment (peterhusisian) [inline cora/generate_svg.html:29]: Commented-out code meant to be uncommented to change functionality can be a nightmare if someone forgets to re-comment it or misses a part, etc.

File: evelyn/misc/precinct_state_wrappers.py (hunk 17 above)
Comment (peterhusisian) [inline evelyn/misc/precinct_state_wrappers.py:1]: Dear lord a 1.3k line file is almost certainly bad practice :|. Is there not some OOP inheritence principles that could have been used to help define wrappers for every state less painful (idk if there are certain fields that can be auto-filled, or perhaps you create an extension to StatePrecinctWrapper for commonly-used subsets of fields, then extend that?). If you're going to make a class for every state, ideally each would be incredibly short for the sake of your, and others', sanity.

File: evelyn/misc/precincts.py [hunk 35]
```diff
@@ -0,0 +1,213 @@
...
+from scipy.stats import t
+import numpy as np
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+from shapely.errors import TopologicalError
+from scipy.spatial.distance import cdist
+
+
+class StatePrecinctWrapper:
```
Comment (peterhusisian) [inline evelyn/misc/precincts.py:15]: What do this do tho? The class name doesn't clarify enough to warrant not explaining what it does.

File: evelyn/misc/precincts.py [hunk 36]
```diff
@@ -0,0 +1,213 @@
...
+
+    def data_report(self):
+        precincts = self.load_precincts()
+        for gdf in precincts:
+            for column in gdf.columns:
+                print(column, (gdf[column] == 0).sum())
+            coverage, shares = self.compute_tract_results(gdf)
+            shares.isna()
+
+    def check_state_mean_and_std(self):
```
Comment (peterhusisian) [inline evelyn/misc/precincts.py:182]: What does "checking" them constitute?

File: evelyn/misc/precincts.py [hunk 37]
```diff
@@ -0,0 +1,213 @@
...
+        for precinct_gdf in precincts:
+            results.update(precinct_gdf.sum(axis=0).to_dict())
+        columns = self.election_columns(include_party=False)
+        election_shares = {c: results['R_' + c] / (results['D_' + c] + results['R_' + c])
+                           for c in columns if 'R_' + c in results}
+        shares = np.array(list(election_shares.values()))
+        print(election_shares)
+        print('Mean', round(shares.mean(), 4), 'STD', round(shares.std(ddof=1), 4))
+
+    def test_self(self):
```
Comment (peterhusisian) [inline evelyn/misc/precincts.py:194]: Makes more sense to test outside of the class, especially since all fields and functions are public

File: evelyn/misc/precincts.py [hunk 38]
```diff
@@ -0,0 +1,213 @@
...
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+from shapely.errors import TopologicalError
+from scipy.spatial.distance import cdist
+
+
+class StatePrecinctWrapper:
+    def __init__(self):
+        self.state = None
```
Comment (peterhusisian) [inline evelyn/misc/precincts.py:17]: All fields and functions in this class are public, which is usually not intended behavior. Fields could be explicitly edited by outside classes which I assume is unintended behavior.
